nixos/modules/services/monitoring/watchdogd.nix

```diff
@@ -0,0 +1,113 @@
+
+  mkPluginOpts = plugin: {
+    enable = mkEnableOption "watchdogd plugin ${plugin}";
+    interval = mkOption {
+      type = types.ints.unsigned;
+      default = 300;
+      description = ''
+        Amount of seconds between every poll.
+      '';
+    };
+    logmark = mkOption {
+      type = types.bool;
+      default = false;
+      description = ''
+        Whether to log current stats every poll interval.
+      '';
+    };
+    warning = mkOption {
+      type = types.ints.unsigned;
+      description = ''
+        The high watermark level. Alert sent to log.
+      '';
+    };
+    critical = mkOption {
+      type = types.ints.unsigned;
+      description = ''
+        The critical watermark level. Alert sent to log, followed by reboot or script action.
+      '';
+    };
+  };
+in {
+  options.services.watchdogd = {
+    enable = mkEnableOption "Advanced system & process supervisor";
+    package = mkPackageOptionMD pkgs "watchdogd" { };
+
+    timeout = mkOption {
+      type = types.ints.unsigned;
+      default = 15;
+      description = ''
+        The WDT timeout before reset.
+      '';
+    };
+    interval = mkOption {
+      type = types.ints.unsigned;
+      default = 5;
+      description = ''
+        The kick interval, i.e. how often {manpage}`watchdogd(8)` should reset the WDT timer.
+      '';
+    };
+
+    safeExit = mkOption {
+      type = types.bool;
+      default = true;
+      description = ''
+        With {var}`safeExit` enabled, the daemon will ask the driver to disable the WDT before exiting.
+        However, some WDT drivers (or hardware) may not support this.
+      '';
+    };
+
+    extraConfig = mkOption {
+      type = types.lines;
+      default = "";
+      description = ''
+        Additional config to put in {file}`watchdogd.conf`.
+        See {manpage}`watchdogd.conf(5)` for syntax.
+     '';
+    };
+
+    filenr = mkPluginOpts "filenr";
+    loadavg = mkPluginOpts "loadavg";
+    meminfo = mkPluginOpts "meminfo";
+  };
+
+  config = let
+    mkPluginConf = plugin:
+      let pcfg = cfg.${plugin};
+      in optionalString pcfg.enable ''
+        ${plugin} {
+          enabled  = true
+          interval = ${toString pcfg.interval}
+          logmark  = ${toString pcfg.logmark}
+          warning  = ${toString pcfg.warning}
+          critical = ${toString pcfg.critical}
+        }
+    '';
+    watchdogdConf = pkgs.writeText "watchdogd.conf" ''
+      timeout = ${toString cfg.timeout}
+      interval = ${toString cfg.interval}
+      safe-exit = ${boolToString cfg.safeExit}
+      ${mkPluginConf "filenr"}
+      ${mkPluginConf "loadavg"}
+      ${mkPluginConf "meminfo"}
+      ${extraConfig}
+    '';
+  in mkIf cfg.enable {
+    environment.systemPackages = [ cfg.package ];
+
+    # Create the service.
+    systemd.services.watchdogd = {
+      wantedBy = [ "multi-user.target" ];
+      description = "Advanced system & process supervisor";
+      restartTriggers = [ watchdogdConf ];
```

Comment:
I do think nixos-rebuild will restart the process whenever the unit changes, and the unit changes anytime this file changes due to the command in `ExecStart`. So this is probably superfluous. Probably not a big deal though.